Sources/CassandraClient/Data+PaginatedRows.swift

```diff
@@ -0,0 +1,245 @@
+        public func nextPage() -> EventLoopFuture<Rows> {
+            guard let eventLoop = self.eventLoop else {
+                preconditionFailure("EventLoop must not be nil")
+            }
+            guard self.hasMorePages else {
+                return eventLoop.makeFailedFuture(CassandraClient.Error.rowsExhausted)
+            }
+
+            let future = self.session.execute(statement: self.statement, on: eventLoop, logger: self.logger)
+            future.whenComplete { result in
+                switch result {
+                case .success(let rows):
+                    self.hasMorePages = cass_result_has_more_pages(rows.rawPointer) == cass_true
+                    if self.hasMorePages {
+                        cass_statement_set_paging_state(self.statement.rawPointer, rows.rawPointer)
+                    }
+                case .failure:
+                    self.hasMorePages = false
+                }
+            }
+            return future
+        }
+
+        /// Iterates through all rows in all pages and invokes the given closure on each.
+        @available(*, deprecated, message: "Use Swift Concurrency and AsyncSequence APIs instead.")
+        public func forEach(_ body: @escaping (Row) throws -> Void) -> EventLoopFuture<Void> {
+            precondition(self.hasMorePages, "Only one of 'forEach' or 'map' can be called once per PaginatedRows")
+
+            guard let eventLoop = self.eventLoop else {
+                preconditionFailure("EventLoop must not be nil")
+            }
+
+            func _forEach() -> EventLoopFuture<Void> {
+                self.nextPage().flatMap { rows in
+                    do {
+                        try rows.forEach(body)
+                    } catch {
+                        return eventLoop.makeFailedFuture(error)
+                    }
+
+                    guard self.hasMorePages else {
+                        return eventLoop.makeSucceededFuture(())
+                    }
+                    return _forEach()
+                }
+            }
+            return _forEach()
+        }
+
+        /// Iterates through all rows in all pages and applies `transform` on each.
+        @available(*, deprecated, message: "Use Swift Concurrency and AsyncSequence APIs instead.")
+        public func map<T>(_ transform: @escaping (Row) throws -> T) -> EventLoopFuture<[T]> {
+            precondition(self.hasMorePages, "Only one of 'forEach' or 'map' can be called once per PaginatedRows")
+
+            guard let eventLoop = self.eventLoop else {
+                preconditionFailure("EventLoop must not be nil in EventLoop based APIs")
+            }
+
+            func _map(_ accumulated: [T]) -> EventLoopFuture<[T]> {
+                self.nextPage().flatMap { rows in
+                    do {
+                        let transformed = try rows.map(transform)
+                        let newAccumulated = accumulated + transformed
```

Comment:
Since accumulated + transformed makes a copy of `accumulated`, and then calls itself (which also makes a copy of all elements of a `newAccumulated`, this algorithm is O(N²).